### PR TITLE
instance args

### PR DESCRIPTION
**Pull request by jmchapman**
- [X] (first attempt at) converting instances args to constraints in type sigs
- [x] not expanding instance usage in clause bodies
- [x] type classes
- [ ] instances
- [ ] Dealing with typically derived stuff like Show

### CODE REVIEW

File: Main.hs [hunk 1]
```diff
@@ -225,7 +230,7 @@ compileTerm builtins v =
     Var x es   -> (`app` es) . Hs.Var () . Hs.UnQual () . hsName =<< showTCM (Var x [])
     Def f es   -> (`app` es) . Hs.Var () =<< hsQName builtins f
     Con h i es -> (`app` es) . Hs.Con () =<< hsQName builtins (conName h)
-    Lit (LitNat n) -> return $ Hs.Lit () $ Hs.Int () n (show n)
+    Lit (LitNat _ n) -> return $ Hs.Lit () $ Hs.Int () n (show n)
```
Comment (UlfNorell) [inline Main.hs:233]: This is a change since `2.6.1`. Probably makes most sense to add a `< 2.6.2` constraint to the cabal file.

Comment (jmchapman) [reply Main.hs:233]: ok, I pushed this cabal file change straight to master.

**Comment by UlfNorell** [comment]
Try a combination of `telViewUpTo`, `recPars`, `defType` to get the parameter telescope. Check what I do in `compileClause` to get it into the context.

**Comment by jmchapman** [comment]
Thanks, I'll give that shot!

**Comment by UlfNorell** [comment]
Would it make sense to limit the scope of this PR and try to get it merged in its current stage (pending merge conflicts)? Just having the translation of class constraints and proper handling of calls to member functions lets you work with the Prelude type classes.

**Comment by jmchapman** [comment]
Yes, good idea. Apologies for blocking this feature. I am planning to get back to agda2hs tomorrow.

I added some support for compiling record declarations into type classes. I haven't dealt with compiling instances of those records to type class instances or picking `deriving` on datatypes. This could be handled in a separate PR.

**Comment by jmchapman** [comment]
I did a merge. The things I have done should probably be made more robust/flexible but perhaps it's already useful?

**Review by UlfNorell — APPROVED** [review]

**Pull request merged**